E2E: Fix gutenberg publish flakiness

lsl opened the pull request:
Fixes #51947 
Fixes #51948 
Fixes #52004 
Fixes #51992 
Fixes #52255

#### Changes proposed in this Pull Request

Address a bunch of issues caused by a flaky "publish" flow of the Gutenberg component.
- Addressed the issue where Payment button causes content change so the page needs to be updated right after being published (https://github.com/Automattic/wp-calypso/issues/50302)
- Made sure that the published page is ensured to be viewed when the `visit` prop is passed
- Removed obsolete `closePanel` prop. No need to close the post-publish panel if we just want to go to the published page/post - the link is in that panel. The panel will be closed, however, if we choose not to visit the published page :)
- Created the `waitUntilElementStopsMoving` helper. It is utilized via the `clickWhenClickable` helper, which should now ensure that we're not attempting to click moving elements. This should generally improve the flakiness caused by interacting with animated elements. The follow-up to this might be https://github.com/Automattic/wp-calypso/issues/52326. 
- Tweaked the revert-to-draft test where an obsolete assertion was used
- Removed unnecessary _waits_ and _sleeps_ 😴 


#### Testing instructions

The following E2E suites should pass:
- Calypso / WPCom Plugins / Gutenberg tests (desktop)
- Calypso / Web app / E2E tests (desktop)
- Calypso / Web app / E2E tests (mobile)

lsl committed:
Fix gutenberg publish flakiness

WunderBart commented on test/e2e/lib/gutenberg/gutenberg-editor-component.js:
```diff
@@ -65,11 +65,17 @@ export default class GutenbergEditorComponent extends AsyncBaseContainer {
+
+		// Clicking publish too soon often fails, I suspect this is due to the slide out animation.
+		// A short sleep is enough to prevent these failures
+		await this.driver.sleep( 10 );
```
Could you link to a job where this is failing, please? It would help ensure we're fixing the right issue here.
As for the fix - 10 milliseconds sound risky to me. Ideally, if the animation is indeed the issue here, we should find out the actual duration of it and use that value for the sleep.

lsl replied:
I'll see if I can trigger it on this one. Was just testing locally and seeing this problem 30% ~ of the time.

lsl replied:
The 10ms - I suspect its something like a next tick timing issue. 1ms wasn't enough but 10ms was, I could try something in-between. I don't know if the transition itself was the problem. I'm thinking a react re/render or something like that.

lsl replied:
[Similar to](https://github.com/Automattic/wp-calypso/pull/52090#discussion_r617143564) here is the failure, happened on my third run:
```
e2e trunk NODE_CONFIG_ENV=decrypted BROWSERSIZE=desktop ./node_modules/.bin/mocha specs/wp-likes-spec.js


  [WPCOM] Likes: (desktop)
    Like posts and comments @parallel
Logging in as e2eflowtestinggutenbergsimple
Publish panel already closed
      1) Login, create a new post and view it
FAILED: Taking screenshot of: 'https://wordpress.com/post/e2eflowtestinggutenbergsimple.wordpress.co/m'
      - Like post
      - Unlike post
      - Post comment
      - Like comment
      - Unlike comment
      - Like post as logged out user


  0 passing (1m)
  6 pending
  1 failing

  1) [WPCOM] Likes: (desktop)
       Like posts and comments @parallel
         Login, create a new post and view it:
     TimeoutError: Waiting for the element to become located and visible By(css selector, .components-snackbar)
Wait timed out after 22043ms
      at /home/louis/wp/wp-calypso/node_modules/selenium-webdriver/lib/webdriver.js:842:17
      at processTicksAndRejections (internal/process/task_queues.js:93:5)



Accepting an open alert: 'UnexpectedAlertOpenError: unexpected alert open: {Alert text : }
  (Session info: chrome=90.0.4430.72)'
(node:167621) UnhandledPromiseRejectionWarning: NoSuchSessionError: This driver instance does not have a valid session ID (did you call WebDriver.quit()?) and may no longer be used.
    at /home/louis/wp/wp-calypso/node_modules/selenium-webdriver/lib/webdriver.js:730:38
    at Object.thenFinally [as finally] (/home/louis/wp/wp-calypso/node_modules/selenium-webdriver/lib/promise.js:124:12)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:167621) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
```

The test run loads a little slow, then when it enters the page title / text you see a quick flash where the sidebar should extend but then nothing and you wait for the timeout.

Adding the spinner displayed check you see the failure far sooner (you don't have to wait for the 60second publish timeout) but the cause is the same - the publish sidebar is never shown. (this is why I wanted to remove the 60 second timeout, seems like this issue would create a lot of false positives for an outlier publish time) 

Or at least, it is but publish is clicked too fast and the sidebar dismisses before the publish can actually happen.

It looks like the publish button closes the sidebar regardless but maybe the publish event isn't wired up until the sidebar is rendered (early maybe?) - I'll take a look over at Gutenberg and see if anything can explain this but I think its worth just dialing in a lower ms sleep and calling it solved.

lsl replied:
Looks like there is a few things that could lead to this sort of thing.

The [actions panel](https://github.com/WordPress/gutenberg/blob/4c81e6e215bf81779920ea7110ca4cacc8bea284/packages/edit-post/src/components/layout/actions-panel.js#L56) loads a [`PostPublishPanel`](https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/editor/src/components/post-publish-panel/index.js#L29). The actons panel has [hints at some unmounting logic](https://github.com/WordPress/gutenberg/blob/4c81e6e215bf81779920ea7110ca4cacc8bea284/packages/edit-post/src/components/layout/actions-panel.js#L51) that could be the problem.

But it could also be something to do with the way we get [`hasPublishAction`](https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/editor/src/components/post-publish-panel/index.js#L146) and then [close if we don't get anything.](https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/editor/src/components/post-publish-panel/index.js#L50).

lsl replied:
Did some more testing with this. The failures drop off after around 3ms but I also managed to trigger it at 10ms so that theory is on shaky ground.

I've decided to bump this up to 100ms to match the [transition time](https://github.com/WordPress/gutenberg/blob/5c78bb872f754b52a3320ed886970dbc0583169e/packages/edit-post/src/components/layout/style.scss#L40) of the animation but I think we probably need to track the failure / retry rate of the test over time to know if we've really nailed it.

lsl replied:
Failure after change mentioned below now looks like this:
```
  1) [WPCOM] Likes: (desktop)
       Like posts and comments @parallel
         Login, create a new post and view it:
     TimeoutError: Waiting for the element to become located and visible By(css selector, .editor-post-publish-panel__header button[aria-label="Close panel"])
Wait timed out after 22244ms
      at /home/louis/wp/wp-calypso/node_modules/selenium-webdriver/lib/webdriver.js:842:17
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

WunderBart replied:
Thanks for digging into this, @lsl! 🙇 I've tried another approach here - instead of _sleeps_, I'm checking if the element is still moving before clicking it (https://github.com/Automattic/wp-calypso/pull/52089/commits/7b32124a4206811202c5cadf284b038402255e21). Including that in the `clickWhenClickable` helper should fix all the issues with moving "unclickable" elements (https://github.com/Automattic/wp-calypso/pull/52089/commits/abd6dd52c4e89068ffaf5861da2a9b0577cb6ca9). I did a quick scan for places that use sleep before clicking elements and removed those _sleeps_ as they should be obsolete now (https://github.com/Automattic/wp-calypso/pull/52089/commits/29ed06261a7613b23287c69adc3d70c74e31ff77). 🤞 In my local testing, I'm able to consistently get pass with this approach.

lsl replied:
Nice work, much better than sprinkling sleeps everywhere.

WunderBart commented on test/e2e/lib/gutenberg/gutenberg-editor-component.js:
```diff
@@ -65,11 +65,17 @@ export default class GutenbergEditorComponent extends AsyncBaseContainer {
-		// Let's give publishing request enough time to finish. Sometimes it takes
-		// way more than the default 20 seconds, and the cost of waiting a bit
-		// longer is definitely lower than the cost of repeating the whole spec.
-		await driverHelper.waitTillNotPresent( this.driver, this.publishingSpinnerSelector, 60000 );
```
The 60 seconds timeout [has been recently added](https://github.com/Automattic/wp-calypso/pull/51575) to address issues where the publish callback was taking longer than usual causing unnecessary retries. Let's bring it back.

lsl replied:
I'm a little concerned about this one, how often does this happen? don't we want to know about it? 60 seconds sounds like a severe problem with the publish api.

lsl replied:
fyi #51863 replaced this line anyway

WunderBart commented on test/e2e/lib/gutenberg/gutenberg-editor-component.js:
```diff
@@ -65,11 +65,17 @@ export default class GutenbergEditorComponent extends AsyncBaseContainer {
+
+		// Ensure the spinner is seen before we confirm it is no longer present.
+		// Without this check, we see "Publish panel already closed" errors when the above misfire occurs.
+		await driverHelper.waitUntilLocatedAndVisible( this.driver, this.publishingSpinnerSelector );
```
👍

lsl committed:
Bump sleep to 100ms matching animation

WunderBart committed:
Add waitUntilElementStopsMoving helper

WunderBart committed:
Use waitUntilElementStopsMoving on publish button

WunderBart committed:
Validate clickable element with waitUntilElementStopsMoving
This should resolve all the moving element issues without any extra sleeps.

WunderBart committed:
Remove unnecessary sleeps before clickable element

lsl commented:
^^ Rebase / conflict fix

lsl commented:
Using 
```
	const commentArea = await CommentsAreaComponent.Expect( driver );
	await commentArea.reply( comment );
```

(with a `console.log(e)` on the try catch in `waitUntilElementStopsMoving`
Generates these:

```
NoSuchElementError: no such element: Unable to locate element: {"method":"css selector","selector":".comment-reply-link"}
  (Session info: chrome=90.0.4430.85)
    at Object.throwDecodedError (/home/louis/wp/wp-calypso/node_modules/selenium-webdriver/lib/error.js:550:15)
    at parseHttpResponse (/home/louis/wp/wp-calypso/node_modules/selenium-webdriver/lib/http.js:565:13)
    at Executor.execute (/home/louis/wp/wp-calypso/node_modules/selenium-webdriver/lib/http.js:491:26)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Driver.execute (/home/louis/wp/wp-calypso/node_modules/selenium-webdriver/lib/webdriver.js:700:17)
    at async /home/louis/wp/wp-calypso/test/e2e/lib/driver-helper.js:619:21 {
  remoteStacktrace: '#0 0x565468ea92b9 <unknown>\n'
}
```

Using 
```
	const commentArea = await CommentsAreaComponent.Expect( driver );
	await commentArea._postComment( comment );
```

Generates these:
```
  ElementNotInteractableError: element not interactable: element has zero size
  (Session info: chrome=90.0.4430.85)
      at Object.throwDecodedError (/home/louis/wp/wp-calypso/node_modules/selenium-webdriver/lib/error.js:550:15)
      at parseHttpResponse (/home/louis/wp/wp-calypso/node_modules/selenium-webdriver/lib/http.js:565:13)
      at Executor.execute (/home/louis/wp/wp-calypso/node_modules/selenium-webdriver/lib/http.js:491:26)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
      at async Driver.execute (/home/louis/wp/wp-calypso/node_modules/selenium-webdriver/lib/webdriver.js:700:17)
      at async Proxy.clickWhenClickable (lib/driver-helper.js:81:3)
      at async CommentsAreaComponent._postComment (lib/pages/frontend/comments-area-component.js:29:3)
      at async Context.<anonymous> (specs/wp-likes-spec.js:69:4)
```

I think this is what you were running into eod yesterday - it might have something to do with seedlet theme, hides the post comment button by default until you click into the comment area.

lsl commented:
Odd that the build is passing - this failure was 100% repeatable locally. I wonder if its a chrome v90 only problem?

(I'll be able to focus on bumping chrome to v90 in e2e tests tomorrow)

WunderBart commented:
> Odd that the build is passing - this failure was 100% repeatable locally. I wonder if its a chrome v90 only problem?

I wasn't able to repro with v90. I tried with the `wp-likes-spec.js` and `wp-comments-spec.js` suites & they're all good and 100% passing. Any idea why we're not getting the same result?

WunderBart committed:
Fix post/page publish flakiness
- Removed unnecessary waits and sleeps
- Addressed the issue where Payment button causes content change so the page needs to be updated right after being published
- Made sure that the published page is actually viewed when "visit" prop is passed
- Removed obsolete "closePanel" prop. No need to close the post-publish panel if we just want to go to the published page/post - the link is in that panel.

lsl committed:
Fix comment posting on desktop screensizes

lsl committed:
No longer required

lsl committed:
Probably still want this too

lsl commented:
@WunderBart it was something to do with window size. Seedlet theme scrolls down when you enter the comment box. I suspect now that we wait for the button to stop moving before it is clicked what happens is the button is pushed down below the fold and then it is clicked but at that point its no longer in view so it errors. Previously we must have been clicking it mid flight so only a single scroll to was enough. I've added a second scroll that happens after the comment has been placed (this is what expands the comment box)

WunderBart commented:
> @WunderBart it was something to do with window size. Seedlet theme scrolls down when you enter the comment box. I suspect now that we wait for the button to stop moving before it is clicked what happens is the button is pushed down below the fold and then it is clicked but at that point its no longer in view so it errors. Previously we must have been clicking it mid flight so only a single scroll to was enough. I've added a second scroll that happens after the comment has been placed (this is what expands the comment box)

Ah, makes sense! Thanks for elaborating. I think that's even better now since we're not trying to click mid-flight? 🥏

WunderBart committed:
Fix failing revert-to-draft test
Removed unnecessary assert because the test would never even get to it. Waiting for the "Revert to draft" button to not be present behaves like assertion: if the button doesn't disappear the test will fail.

WunderBart commented:
All checks have passed! 🎉 Going to merge this one.
